### PR TITLE
Rebased better-merge-exception

### LINKED ISSUE #373: When you try to merge a branch that isn't mergeable, get a better exception
As per http://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button the API returns a `405 Method not allowed` when the PR is set as `mergeable: false`

We should raise a specific exception here, rather than the generic API exception...

### PR DESCRIPTION
**Pull request by elbaloo**
Rebased @shiftkey's from #819 and mine from #931 on top of master.
Fixes #373.

### CODE REVIEW

File: Octokit/Exceptions/PullRequestNotMergeableException.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+namespace Octokit
+{
+    /// <summary>
+    /// Represents an error that occurs when the pull request is in an
+    /// unmergeable state
+    /// </summary>
+#if !NETFX_CORE
+    [Serializable]
+#endif
+    public class PullRequestNotMergeableException : Exception
```
Comment (shiftkey) [inline Octokit/Exceptions/PullRequestNotMergeableException.cs]: via @haacked from some other PR:

> Yeah, they probably should inherit from `ApiException`.

File: Octokit/Exceptions/PullRequestMismatchException.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+namespace Octokit
+{
+    /// <summary>
+    /// Represents an error that occurs when the specified SHA
+    /// doesn't match the current pull request's HEAD
+    /// </summary>
+#if !NETFX_CORE
+    [Serializable]
+#endif
+    public class PullRequestMismatchException : Exception
```
Comment (shiftkey) [inline Octokit/Exceptions/PullRequestMismatchException.cs]: via @haacked from some other PR:

> Yeah, they probably should inherit from `ApiException`.

**Comment by elbaloo** [comment]
@Haacked / @shiftkey let me know if this looks ok.
btw, I changed the messages to be those that the github api returns.

**Comment by shiftkey** [comment]
@elbaloo apologies for the delay - this looks great! It just needs a merge from `master` to address the csproj conflicts...

**Comment by Haacked** [comment]
Closed by #976

**Pull request closed without merging**